### PR TITLE
DROOLS-4407: [DMN Designer] Search feature - Highlight the current result with colors that follow the PF standard

### PR DESCRIPTION
**Pull request by danielzhe**
This PR is part of an ensemble and also covers [DROOLS-4406](https://issues.jboss.org/projects/DROOLS/issues/DROOLS-4406).

https://github.com/kiegroup/drools-wb/pull/1214
https://github.com/kiegroup/appformer/pull/767
https://github.com/kiegroup/kie-wb-common/pull/2815

### CODE REVIEW

File: kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/search/common/BaseEditorSearchIndex.java [hunk 1]
```diff
@@ -183,7 +195,13 @@ private void triggerNoResultsFoundCommand() {
...
     }
 
-    private Optional<T> getCurrentResult() {
+    void triggerOnSearchPerformedCommand() {
+        if (!Objects.isNull(searchPerformedCallback)) {
+            searchPerformedCallback.execute();
+        }
+    }
+
+    public Optional<T> getCurrentResult() {
```
Comment (jomarko) [inline kie-wb-common-widgets/kie-wb-common-ui/src/main/java/org/kie/workbench/common/widgets/client/search/common/BaseEditorSearchIndex.java]: Please add `@Override` annotation

**Comment by danielzhe** [comment]
@karreiro I think it's fixed now by changing the offset. I don't recall why I was using -offset. I did some more tests now and didn't found any issue using +offset.

**Comment by danielzhe** [comment]
@karreiro Sorry, commented in the wrong PR. I'm still working in this one. The PR related to search in GDT is the one fixed.

**Comment by danielzhe** [comment]
@karreiro Changes applied.

**Comment by jomarko** [comment]
@danielzhe we have a checkstyle violation, please fix it:
```
[ERROR] src/main/java/org/kie/workbench/common/dmn/client/editors/search/DMNEditorSearchIndex.java:[21] (imports) CustomImportOrder: Wrong lexicographical order for 'java.util.Optional' import. Should be before 'java.util.function.Supplier'.
```

**Comment by jomarko** [comment]
@danielzhe I think some incompatible changes were merged meanwhile. I resolved checksyle violation to proceed locally, but getting:
```
[ERROR] /home/jomarko/redhat/github/kiegroup/kie-wb-common/kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/editors/search/DMNEditorSearchIndexTest.java:[67,27] constructor DMNEditorSearchIndex in class org.kie.workbench.common.dmn.client.editors.search.DMNEditorSearchIndex cannot be applied to given types;
[ERROR]   required: org.kie.workbench.common.dmn.client.editors.search.DMNGraphSubIndex,org.kie.workbench.common.dmn.client.editors.search.DMNGridSubIndex,org.kie.workbench.common.dmn.client.editors.search.DMNDataTypesSubIndex,org.kie.workbench.common.stunner.core.client.api.SessionManager,javax.enterprise.event.Event<org.kie.workbench.common.widgets.client.search.common.SearchPerformedEvent>
[ERROR]   found: org.kie.workbench.common.dmn.client.editors.search.DMNGraphSubIndex,org.kie.workbench.common.dmn.client.editors.search.DMNGridSubIndex,org.kie.workbench.common.dmn.client.editors.search.DMNDataTypesSubIndex,org.kie.workbench.common.stunner.core.client.api.SessionManager
[ERROR]   reason: actual and formal argument lists differ in length
```

Could you please check if your ensemble can be compiled on your machine?

**Comment by danielzhe** [comment]
@jomarko Sorry, I forgot one commit. It's done now.

**Comment by danielzhe** [comment]
Nice, more conflicts :(

**Comment by jomarko** [comment]
@danielzhe these seems to be due to merge of my PR #2797 . Let me know if you need help with resolving conflicts.

**Comment by danielzhe** [comment]
@jomarko Merged.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java [hunk 2]
```diff
@@ -194,6 +194,24 @@ public void setExpression(final String nodeUUID,
         return existingWidth;
     }
 
+    Optional<BaseExpressionGrid> getExistingEditor() {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionContainerGrid.java]: I think we now need to reuse this also in the `getExistingWidth`, my proposal:
```java
    double getExistingEditorWidth() {
        double existingWidth = DMNGridColumn.DEFAULT_WIDTH;
        if (getExistingEditor().isPresent()) {
            final BaseExpressionGrid beg = getExistingEditor().get();
            existingWidth = Collections.max(Arrays.asList(existingWidth,
                                                          beg.getWidth() + beg.getPadding() * 2,
                                                          beg.getMinimumWidth() + beg.getPadding() * 2));
        }
        return existingWidth;
    }

    Optional<BaseExpressionGrid<? extends Expression, ? extends GridData, ? extends BaseUIModelMapper>> getExistingEditor() {

        final GridCell<?> cell = model.getRow(0).getCells().get(0);
        if (cell != null) {
            final GridCellValue<?> value = cell.getValue();
            if (value instanceof ExpressionCellValue) {
                final ExpressionCellValue ecv = (ExpressionCellValue) value;
                return ecv.getValue();
            }
        }

        return Optional.empty();
    }
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorViewImpl.java [hunk 3]
```diff
@@ -131,6 +140,30 @@ public ExpressionEditorViewImpl(final Anchor returnToLink,
         this.domainObjectSelectionEvent = domainObjectSelectionEvent;
     }
 
+    void onSearchPerformed(final @Observes SearchPerformedEvent event) {
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/ExpressionEditorViewImpl.java]: My previous comment would require changes also here:
```java
    void onSearchPerformed(final @Observes SearchPerformedEvent event) {
        getExistingEditor().ifPresent(e -> {
            final GridRenderer renderer = e.getRenderer();
            if (renderer instanceof BaseGridRenderer) {
                final BaseGridRenderer bgr = (BaseGridRenderer) renderer;
                bgr.clearCellHighlight();
                if (event.hasElement()) {
                    final Searchable element = event.getCurrentElement();
                    if (element instanceof DMNSearchableElement) {
                        final DMNSearchableElement gdtElement = ((DMNSearchableElement) element);
                        int column = gdtElement.getColumn();
                        int row = gdtElement.getRow();
                        bgr.highlightCell(column, row);
                    }
                }
            }
        });
    }

    Optional<BaseExpressionGrid<? extends Expression, ? extends GridData, ? extends BaseUIModelMapper>> getExistingEditor() {
        return expressionContainerGrid.getExistingEditor();
    }
```

**Comment by jomarko** [comment]
### Manual review note 1
Sees highlight doesn't work for all expression types, see the [video](https://issues.jboss.org/secure/attachment/12454528/search-context.webm), where I search **context** expression from the sample **traffic violation** project.

**Comment by jomarko** [comment]
### Manual review note 2
Merged view of guided decision tables doesn't highlight properly.
Screenshot from 2019-08-21 09-57-14

**Comment by jomarko** [comment]
### Manual review note 3
We need to differentiate closing search bar component by clicking **cross** and **magnifier** . With your changes we clear selection completely in both cases, what makes grid inaccessible by keyboard, what is an issue for selenium tests. Could we please change behavior to keep the cell selected after search bar component is closed? At leas for one case, closing by **cross** or **magnifier**.

**Comment by danielzhe** [comment]
@jomarko I made some changes to simplify the code. Since the approach changed a bit some code are also not required anymore and I think all reported issues are fixed now.

**Comment by jomarko** [comment]
### Manual review, round 2, notes
- The guided decision table search highlight now works well, even for merged cells thank you. There is just the need of keyboard control for selenium tests, I reported an issue for myself https://issues.jboss.org/browse/DROOLS-4456
- The dmn expression highlight changed behavior, but still does not work well
  - See the video https://issues.jboss.org/secure/attachment/12454571/search-context-2.webm
  - After closing search in DMN expression, the highlight remains in place
  - Screenshot from 2019-08-22 07-41-11
  - Screenshot from 2019-08-22 07-41-18
  - The highlight of dmn expression remains even if you edit some other expression in the meantime, not sure, is this intentional?
- The dmn expressions keyboard control works after search is closed

**Comment by danielzhe** [comment]
@jomarko I didn't incorporate because of time issues trying to merge this PR before Friday.

**Comment by danielzhe** [comment]
@jomarko It's fixed now.

**Comment by jomarko** [comment]
@danielzhe I can confirm it is fixed now, however other two parts of this ensemble have some comments form me where I am waiting for your answer.

**Comment by danielzhe** [comment]
@jomarko Thank you. I was waiting to see if there is any issue before proceed to the other changes that you proposed. I'll let you know when is done. Thank you.

**Pull request merged**